Pat_SO3.cpp

```diff
@@ -0,0 +1,66 @@
+#include "Pat_SO3.h"
+#include <iostream>
+#include <cmath>
+#include "Eigen/Dense"
+
+Pat_SO3::Pat_SO3(Eigen::Matrix3f R) {
+    if (is_SO3(R)) {
+        _rot = R;
+    } else {
+        _rot = Eigen::Matrix3f::Identity();
+    }
+}
+
+bool Pat_SO3::is_SO3(const Eigen::Matrix3f& R) {
+    return true;  // Always true for 3x3
+}
+
+
+Pat_SO3 Pat_SO3::from_matrix(const Eigen::Matrix3f& R) {
+    if (is_SO3(R)) {
+        return Pat_SO3(R);
+    } else {
+        throw std::invalid_argument("Not a valid SO3 matrix");
+    }
+}
+
+Eigen::Matrix3f Pat_SO3::skew(const Eigen::Vector3f& v) {
+    Eigen::Matrix3f skew_matrix;
+    skew_matrix << 0, -v(2), v(1),
+                   v(2), 0, -v(0),
+                  -v(1), v(0), 0;
+    return skew_matrix;
+}
+
+Eigen::Matrix3f Pat_SO3::matrix_exponential(const Eigen::Matrix3f& M) {
```

Comment:
In this current implementation, you would need to check that M is a skew-symmetric matrix. So I would probably accept a vector3f argument instead, and then convert that to a skew symmetric matrix inside the function.

An alternative is to instead define a type for skew-symmtric matrices and then accept that type as the function argument.